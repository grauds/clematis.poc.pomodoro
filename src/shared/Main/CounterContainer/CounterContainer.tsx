import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";

import {
  RootState,
  updateTask,
  updateTaskPomodoro,
} from "../../../store/reducer";
import {
  EPomodoroStatus,
  ETaskStatus,
  IPomodoro,
  ITask,
  freshPomodoro,
} from "../../../types/model";

import { Counter, ICounterProps } from "../Counter/Counter";
import { noop } from "../../../utils/noop";


export function CounterContainer() {

  // current task reference
  const currentTask = useSelector<RootState, ITask | undefined>((state) =>
    state.tasks.length > 0 ? state.tasks[0] : undefined
  );

  // get next pomodoro
  let currentPomodoro: IPomodoro | undefined = currentTask?.pomodori.find(
    (pomodoro: IPomodoro) => {
      return pomodoro.status !== EPomodoroStatus.DONE;
    }
  );

  const seconds = currentPomodoro ? currentPomodoro.seconds : 0;
  const breakSeconds = currentPomodoro ? currentPomodoro.breakSeconds : 0;

  const isTaskRunning = currentPomodoro?.status === EPomodoroStatus.RUNNING;
  const isPause = currentPomodoro?.status === EPomodoroStatus.PAUSED;
  const isBreakRunning = currentPomodoro?.status === EPomodoroStatus.BREAK_RUNNING;

  const dispatch = useDispatch();

  useEffect(() => {
    const timer = setInterval(() => {
      if (currentPomodoro) {
        if (
          (isTaskRunning && seconds > 0) ||
          (isBreakRunning && breakSeconds > 0)
        ) {
          if (isTaskRunning) {
            currentPomodoro.seconds -= 1;
            currentPomodoro.time += 1;
          } else if (isBreakRunning) {
            currentPomodoro.breakSeconds -= 1;
            currentPomodoro.break += 1;
          }
          dispatch(updateTaskPomodoro(currentPomodoro));
        }
        if (isTaskRunning && seconds === 0) {
          currentPomodoro.status = EPomodoroStatus.BREAK_RUNNING;
          dispatch(updateTaskPomodoro(currentPomodoro));
        }
        if (isBreakRunning && breakSeconds === 0) {
          finishPomodoroOrTask() 
        }
        if (isPause) {
          currentPomodoro.pause += 1;
          dispatch(updateTaskPomodoro(currentPomodoro));
        }
      }
    }, 1000);

    return () => clearInterval(timer);
  }, [currentTask]);

  function startTimer() {
    if (currentTask && currentPomodoro) {
      currentPomodoro.status = EPomodoroStatus.RUNNING;
      currentTask.status = ETaskStatus.RUNNING;
      dispatch(updateTaskPomodoro(currentPomodoro));
      dispatch(updateTask(currentTask));
    }
  }

  function pauseTimer() {
    if (currentTask && currentPomodoro) {
      currentPomodoro.status = EPomodoroStatus.PAUSED;
      currentTask.status = ETaskStatus.PAUSED;
      dispatch(updateTaskPomodoro(currentPomodoro));
      dispatch(updateTask(currentTask));
    }
  }

  function pauseBreakTimer() {
    if (currentTask && currentPomodoro) {
      currentPomodoro.status = EPomodoroStatus.BREAK_PAUSED;
      currentTask.status = ETaskStatus.PAUSED;
      dispatch(updateTaskPomodoro(currentPomodoro));
      dispatch(updateTask(currentTask));
    }
  }

  function stopTimer() {
    if (currentTask && currentPomodoro) {
      currentPomodoro = {
        id: currentPomodoro.id,
        ...freshPomodoro,
      };
      dispatch(updateTaskPomodoro(currentPomodoro));
      dispatch(updateTask(currentTask));
    }
  }

  function finishPomodoro() {
    if (currentTask && currentPomodoro) {
      currentPomodoro.status = EPomodoroStatus.DONE;
      currentTask.status = ETaskStatus.PAUSED
      dispatch(updateTaskPomodoro(currentPomodoro));
      dispatch(updateTask(currentTask));
    }
  }

  function startBreakTimer() {
    if (currentTask && currentPomodoro) {
      currentPomodoro.status = EPomodoroStatus.BREAK_RUNNING;
      currentTask.status = ETaskStatus.RUNNING;
      dispatch(updateTaskPomodoro(currentPomodoro));
      dispatch(updateTask(currentTask));
    }
  }

  function finishPomodoroOrTask() {
    if (currentTask && currentPomodoro) {
      currentPomodoro.status = EPomodoroStatus.DONE;
      currentTask.status =
        currentTask.pomodori.length !== currentPomodoro.id
          ? ETaskStatus.PAUSED
          : ETaskStatus.DONE;
      dispatch(updateTaskPomodoro(currentPomodoro));
      dispatch(updateTask(currentTask));
    }
  }


  const empty: ICounterProps =  {
    active: false,
    name: '',
    pomodoroNo: 0,
    titleCss: '',
    seconds: 0,
    secondsCss: "",
    bodyCss: "",
    leftButtonTitle: "",
    leftButtonAction: noop,
    rightButtonTitle: "",
    rightBuittonAction: noop,
    rightButtonDisabled: false,
    handleTimeAdd: noop
  }

  const notStarted: ICounterProps =  {
    active: true,
    name: currentTask ? currentTask.name : 'Без имени',
    pomodoroNo: currentPomodoro ? currentPomodoro.id : 0,
    seconds: currentPomodoro ? currentPomodoro.seconds : 0,
    titleCss: '',
    secondsCss: '',
    bodyCss: '',
    leftButtonTitle: "Старт",
    leftButtonAction: startTimer,
    rightButtonTitle: "Стоп",
    rightBuittonAction: noop,
    rightButtonDisabled: true,
    handleTimeAdd: () => {
      if (currentPomodoro && currentPomodoro.seconds > 0) {
          currentPomodoro.seconds += 60;
          dispatch(updateTaskPomodoro(currentPomodoro));
      }
    }
  }

  const pauseWork: ICounterProps =  {
    ...notStarted,
    titleCss: 'running',
    secondsCss: 'running',
    bodyCss: 'running',
    leftButtonTitle: "Пауза",
    leftButtonAction: pauseTimer,
    rightButtonTitle: "Стоп",
    rightBuittonAction: stopTimer,
    rightButtonDisabled: false
  }

  const afterPause: ICounterProps = {
    ...pauseWork,
    leftButtonTitle: "Продолжить",
    leftButtonAction: startTimer,
    rightButtonTitle: "Сделано",
    rightBuittonAction: finishPomodoroOrTask
  }

  const pauseBreak: ICounterProps =  {
    ...notStarted,
    seconds: currentPomodoro ? currentPomodoro.breakSeconds : 0,
    titleCss: 'breakRunning',
    secondsCss: 'breakRunning',
    bodyCss: 'breakRunning',
    leftButtonTitle: "Пауза",
    leftButtonAction: pauseBreakTimer,
    rightButtonTitle: "Пропустить",
    rightBuittonAction: finishPomodoroOrTask,
    rightButtonDisabled: false,
    handleTimeAdd: () => {
      if (currentPomodoro) {
          currentPomodoro.breakSeconds += 60;
          dispatch(updateTaskPomodoro(currentPomodoro));
      }
    }
  }

  const afterBreakPause: ICounterProps = {
    ...pauseBreak,
    leftButtonTitle: "Продолжить",
    leftButtonAction: startBreakTimer,
  }

  let counterProps: ICounterProps = empty

  switch (currentPomodoro?.status) {
    case EPomodoroStatus.NOT_STARTED:
      counterProps = notStarted
      break;
    case EPomodoroStatus.RUNNING:
      counterProps = pauseWork
      break;
    case EPomodoroStatus.PAUSED:
      counterProps = afterPause
      break;
    case EPomodoroStatus.BREAK_RUNNING:
      counterProps = pauseBreak
      break;
    case EPomodoroStatus.BREAK_PAUSED:
      counterProps = afterBreakPause
      break;
    case EPomodoroStatus.DONE:
      counterProps = empty
      break;   
  }

  return (
     <Counter {...counterProps} />
  );
}
