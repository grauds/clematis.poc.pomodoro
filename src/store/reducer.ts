import { ActionCreator, AnyAction, Reducer } from "redux";
import { Days, IDay, ITask, IWeek, Weeks } from "../types/model";

export type RootState = {
  tasks: ITask[];
  day: IDay;
  week: IWeek;
};

const initialState: RootState = {
  tasks: [],
  day: Days[0],
  week: Weeks[0],
};

const ADD_TASK = "ADD_TASK";
const REMOVE_TASK = "REMOVE_TASK";
const SET_CURRENT_DAY = "SET_CURRENT_DAY";
const SET_CURRENT_WEEK = "SET_CURRENT_WEEK";

export const addTask: ActionCreator<AnyAction> = (task: ITask) => ({
  type: ADD_TASK,
  task,
});

export const removeTask: ActionCreator<AnyAction> = (task: ITask) => ({
  type: REMOVE_TASK,
  task,
});

export const setCurrentDay: ActionCreator<AnyAction> = (day: IDay) => ({
  type: SET_CURRENT_DAY,
  day,
});

export const setCurrentWeek: ActionCreator<AnyAction> = (week: IWeek) => ({
  type: SET_CURRENT_WEEK,
  week,
});

export const rootReducer: Reducer<RootState> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case ADD_TASK:
      return {
        ...state,
        tasks: state.tasks.concat(action.task),
      };
    case REMOVE_TASK:
      return {
        ...state,
        tasks:
          state.tasks.indexOf(action.task) !== -1
            ? state.tasks.splice(action.task, 1)
            : state.tasks,
      };
    case SET_CURRENT_DAY:
      return {
        ...state,
        day: action.day,
      };  
    case SET_CURRENT_WEEK:
      return {
        ...state,
        week: action.week,
      };
    default:
      return state;
  }
};
