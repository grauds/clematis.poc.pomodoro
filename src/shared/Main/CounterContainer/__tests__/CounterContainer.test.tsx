import React from 'react';
import { CounterContainer } from '../CounterContainer';
import { createStore } from 'redux';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';

import { initialState, rootReducer } from '@/store/reducer';

import { render } from '@testing-library/react';

describe('Counter container', () => {
  test('should render', () => {
    const store = createStore(rootReducer, initialState);

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <CounterContainer />
        </MemoryRouter>
      </Provider>,
    );
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });
});
