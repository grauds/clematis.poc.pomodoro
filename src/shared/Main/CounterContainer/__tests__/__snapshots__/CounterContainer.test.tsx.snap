// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter container should render 1`] = `
<div>
  <div
    class="counter"
  >
    <div
      class="counterheader  "
    >
      <h2>
        Задача не выбрана
      </h2>
      <span
        class="total"
      >
        Помидор: 
        0
      </span>
    </div>
    <span
      class="body"
    />
  </div>
</div>
`;
